// ng2-simple-smooth-scroll/esm2015/lib/ng2-simple-smooth-scroll.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, HostListener, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { SimpleSmoothScrollService } from './ng2-simple-smooth-scroll.service';
export class SimpleSmoothScrollDirective {
    /**
     * @param {?} el
     * @param {?} smooth
     * @param {?} platformId
     */
    constructor(el, smooth, platformId) {
        this.el = el;
        this.smooth = smooth;
        this.platformId = platformId;
    }
    /**
     * @return {?}
     */
    onClick() {
        if (isPlatformBrowser(this.platformId)) {
            /** @type {?} */
            let eid = this.el.nativeElement.hash;
            if (eid) {
                if (this.showHash) {
                    // Change URL hash without page jump
                    history.pushState(null, null, eid);
                }
                eid = eid.replace('#', '');
            }
            /** @type {?} */
            let target = document.getElementById(eid) || document.getElementById(this.scrollTo);
            if (target)
                this.smooth.smoothScroll(target.offsetTop, { duration: this.duration, easing: this.easing, offset: this.offset });
            return false;
        }
    }
}
SimpleSmoothScrollDirective.decorators = [
    { type: Directive, args: [{
                selector: '[SimpleSmoothScroll]'
            },] }
];
/** @nocollapse */
SimpleSmoothScrollDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: SimpleSmoothScrollService },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
SimpleSmoothScrollDirective.propDecorators = {
    scrollTo: [{ type: Input, args: ['scrollTo',] }],
    duration: [{ type: Input, args: ['duration',] }],
    offset: [{ type: Input, args: ['offset',] }],
    easing: [{ type: Input, args: ['easing',] }],
    showHash: [{ type: Input, args: ['showHash',] }],
    onClick: [{ type: HostListener, args: ['click',] }]
};
if (false) {
    /** @type {?} */
    SimpleSmoothScrollDirective.prototype.scrollTo;
    /** @type {?} */
    SimpleSmoothScrollDirective.prototype.duration;
    /** @type {?} */
    SimpleSmoothScrollDirective.prototype.offset;
    /** @type {?} */
    SimpleSmoothScrollDirective.prototype.easing;
    /** @type {?} */
    SimpleSmoothScrollDirective.prototype.showHash;
    /** @type {?} */
    SimpleSmoothScrollDirective.prototype.el;
    /** @type {?} */
    SimpleSmoothScrollDirective.prototype.smooth;
    /** @type {?} */
    SimpleSmoothScrollDirective.prototype.platformId;
}
