// angular-mgl-timeline/esm5/src/timeline/timeline-entry-content/timeline-entry-content.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { AnimationBuilder, style, animate } from '@angular/animations';
import { Component, EventEmitter, ElementRef, Renderer } from '@angular/core';
var MglTimelineEntryContentComponent = /** @class */ (function () {
    function MglTimelineEntryContentComponent(elementRef, animationBuilder, renderer) {
        this.elementRef = elementRef;
        this.animationBuilder = animationBuilder;
        this.renderer = renderer;
        this.animationDone = new EventEmitter();
        this._expanded = false;
    }
    Object.defineProperty(MglTimelineEntryContentComponent.prototype, "expanded", {
        get: /**
         * @return {?}
         */
        function () {
            return this._expanded;
        },
        set: /**
         * @param {?} expanded
         * @return {?}
         */
        function (expanded) {
            this.contentHeight = this.elementRef.nativeElement.scrollHeight;
            /** @type {?} */
            var animate = this._expanded !== expanded;
            this._expanded = expanded;
            animate ? this.animate() : this.setStyle;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    MglTimelineEntryContentComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.contentHeight = this.elementRef.nativeElement.scrollHeight;
        this.setStyle();
    };
    /**
     * @return {?}
     */
    MglTimelineEntryContentComponent.prototype.getCollapsedStyle = /**
     * @return {?}
     */
    function () {
        return {
            height: 0
        };
    };
    /**
     * @return {?}
     */
    MglTimelineEntryContentComponent.prototype.getExpandedStyle = /**
     * @return {?}
     */
    function () {
        return {
            height: this.contentHeight + 'px'
        };
    };
    /**
     * @return {?}
     */
    MglTimelineEntryContentComponent.prototype.animate = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.expanded) {
            /** @type {?} */
            var animation = this.animationBuilder
                .build([
                style(this.getCollapsedStyle()),
                animate('100ms ease', style(this.getExpandedStyle())),
            ])
                .create(this.elementRef.nativeElement);
            animation.onDone(function () { return _this.animationDone.emit({ toState: 'expanded' }); });
            animation.play();
        }
        else {
            this.animationBuilder;
            /** @type {?} */
            var animation = this.animationBuilder
                .build([
                style(this.getExpandedStyle()),
                animate('200ms ease', style(this.getCollapsedStyle())),
            ])
                .create(this.elementRef.nativeElement);
            animation.onDone(function () { return _this.animationDone.emit({ toState: 'collapsed' }); });
            animation.play();
        }
    };
    /**
     * @return {?}
     */
    MglTimelineEntryContentComponent.prototype.setStyle = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var baseStyle = this.expanded ? this.getExpandedStyle() : this.getCollapsedStyle();
        Object.keys(baseStyle).forEach(function (property) {
            _this.renderer.setElementStyle(_this.elementRef.nativeElement, property, baseStyle[property]);
        });
    };
    MglTimelineEntryContentComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mgl-timeline-entry-content',
                    template: "<div><ng-content></ng-content></div>",
                    styles: [":host{position:relative;display:block;overflow:hidden}:host>div{padding:10px}"]
                }] }
    ];
    /** @nocollapse */
    MglTimelineEntryContentComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: AnimationBuilder },
        { type: Renderer }
    ]; };
    return MglTimelineEntryContentComponent;
}());
export { MglTimelineEntryContentComponent };
if (false) {
    /** @type {?} */
    MglTimelineEntryContentComponent.prototype.contentHeight;
    /** @type {?} */
    MglTimelineEntryContentComponent.prototype.animationDone;
    /** @type {?} */
    MglTimelineEntryContentComponent.prototype._expanded;
    /** @type {?} */
    MglTimelineEntryContentComponent.prototype.elementRef;
    /** @type {?} */
    MglTimelineEntryContentComponent.prototype.animationBuilder;
    /** @type {?} */
    MglTimelineEntryContentComponent.prototype.renderer;
}
